### PR TITLE
User Permissions

### PR DESCRIPTION
**Pull request by stevezhang1999**
Notice:
- error handling and some other details in the client still need a better design.
- the problem from the last PR, handling the deletion of the projects, is still in progress.

### Description of changes
New features:
- now each project is corresponding to one user.
- permissions
  - usergroup: normal user
    - manipulating files created by himself/herself.
- manage users and other data via Django's admin site. (localhost:port/admin)

Change in implementation:
- the new implementation of authentication and permission system is based on Django's auth module.

### Possible influences of this PR
No negative influence expected.

### Demonstration of results
summary of the example databse
- 1 superuser
  - username _admin_, password _admin_
  - has 2 example projects, each one with 1 document, same as before.
  - has all kinds of access
- 1 normal user
  - username _normal_, password _example123_
  - has 1 example projects, with 1 document.
  - able to add projects and documents, upgrade, delete, read ones created by himself/herself
  - user group implementation detail (accessible via admin site)
    - permissions: "add project" and "add document"

tests:

**normal user**

- valid operations
  - create project :heavy_check_mark:
  - create document for the project :heavy_check_mark:
  - delete the document :heavy_check_mark:
  - delete the project :heavy_check_mark:
  - add annotaion to the document :heavy_check_mark:
  - delete annotation from the document :heavy_check_mark:
  - add link to the document (not done)
  - delete link form the document (not done)
- invalide operations
  - access document not belonged to the user :heavy_check_mark:
  - access project not belonged to the user:heavy_check_mark:

### CODE REVIEW

File: simple-backend/nlpviewer_backend/handlers/document.py [hunk 1]
```diff
@@ -42,7 +71,7 @@ def edit(request, document_id):
     docJson = model_to_dict(doc)
     return JsonResponse(docJson, safe=False)
 
-# need rewrite 
+# DEPRECATED - need rewrite
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/document.py:74]: You can use the word TODO here, many IDE will search for TODO and remind you.

File: simple-backend/nlpviewer_backend/handlers/document.py [hunk 2]
```diff
@@ -32,7 +47,21 @@ def create(request):
...
+    accessible for users with 'change' permission and the owner of the project.
+    """
+    # if doc doen't exist
+    try:
+        doc = Document.objects.get(pk=document_id)
+    except ObjectDoesNotExist: 
+        return HttpResponse('not found', status=404)
+
+    # check permissions
+    if not request.user.has_perm('nlpviewer_backend.change_document') and request.user != doc.project.user:
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/document.py:61]: So the permission condition here is:

1. The user can change document
2. The user belongs to this project

What happen if we want one user to be able to edit one project but not the other one?

**Comment by stevezhang1999** [comment]
> My main question would be on the permission mechanism.
> 
> If there are two projects, a user belongs to both projects. Can we configure so that the users have different roles on these two projects?
> 
> It looks like to me that there is only one user per project? Normally there should be multiple user per project, and each user can have different roles here, at least two different roles: editor or viewer.
> 
> Maybe we can keep the role of the normal user you defined (the superuser for a project, who can delete or add documents). But we should also allow multiple normal users to edit documents (maybe not deleting or adding documents)
> 
> Do you think this can be easily changed in the current PR, or you would like us to merge this one and start another PR for that?

OK I will re-design the permission mechanism. Thanks for your suggestions. I prefer not merge until the completion of this module.

**Comment by stevezhang1999** [comment]
### summary
1. uses additional module， django-guardian
  ```
pip install django-guardian
```

2. now can assign permissions **per project, per user**. currently two permissions, viewer and editor

3. how to use the new API: 

    ```python
      # Or justt do this in the admin site, see below
      assign_perm('viewer_project', joe, project)
      
      joe.has_perm('viewer_project', project)
      ```

### current permission design:

Group "normal user": add_document, add_project

- allow user to create new projects and documents.

To be a viewer: viewer_document

- can view the project and its documents.

To be a editor: editor_document or viewer_document

- can edit the documents, but cannot delete a document or the project.



In the permissions panel there are 6 permissions, including 4 default permissions and 2 additional permissions: "Can view **the** project" and "Can edit the project". The 4 default permissions should be ignored. The "Can view the project" means the user is a viewer and the latter means the user is an editor.



Currently， users excluding staff could have **one and only one** state of the following for each document:

- has the permission "Can view the project" - viewer_project
- has the permission "Can edit the project" - editor_project
- is the owner of the project
- has neither of the two permissions and is not the owner



To assign permission per object:

1. click the project
     



2. click "Object Permissions" in the upper-right corner



3. Enter exact user name to find or edit users already modified before



4. Assign permissions and ignore default ones.





TODO:

- The viewer could still try to add an annotation, which will cause an error. The add annotation button should be disabled for viewers.
- It would be better that the 'projects' page could be separated into 3 sections: My Projects, As Viewers and As  Editors. As a result, the function `list_user_projects` should also be separated into 3 functions.

File: simple-backend/nlpviewer_backend/handlers/document.py [hunk 3]
```diff
@@ -65,23 +128,92 @@ def edit_ontology(request, document_id):
...
+            {"id": 42, "name": "project1-doc1-example", "project_id": 5, "textPack": "...", ...}    
+    """
+    # if doc doen't exist
+    try:
+        doc = Document.objects.get(pk=document_id)
+    except ObjectDoesNotExist: 
+        return HttpResponse('not found', status=404)
+
+    # check permissions
+    if (not request.user.has_perm('nlpviewer_backend.viewer_project', doc.project) and
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/document.py:153]: The permission check has been called many times. I have a suggestion for this, you can create a decorator function for this check `doc_permission(permission_name: str)`, when using this on a function, all you need to do is to put `@permission(some_permission_name)` before each function. For example, in this case, you will simply write `@permission('nlpviewer_backend.viewer_project')`. Now you can let your decorator function to access the parameter of this function, say the `doc_id`, then check if the `doc_id` exists, and check if the permission is OK. You can take a look here for example:

https://stackoverflow.com/questions/1010080/python-get-original-function-arguments-in-decorator

You can also follow the existing decorator such as `require_login`

**Review by hunterhector — COMMENTED** [review]
1. You can use decorator to handle the permission, 
2. Is it possible to do unit tests for these functions? I saw you commenting that a lot of the code are not called. Unit tests can help check the basic logics of such methods.

**Comment by stevezhang1999** [comment]
### summary
1. uses additional module， django-guardian
  ```
pip install django-guardian
```

2. now can assign permissions **per project, per user**. currently 6 per object permissions are available.

3. how to use the new API: 

    ```python
      # Or justt do this in the admin site, see below
      assign_perm('view_project', joe, project)
      
      joe.has_perm('view_project', project)
      ```
4. now most of the comments are written in Google style.
### current permission design:

### per object permissions: 6 permissions
- read_project:
- edit_annotation
- edit_text
- edit_project
- remove_project
- new_project

### default django permissions:
- add_project
- view_project






### To assign permission per object:

1. click the project
     



2. click "Object Permissions" in the upper-right corner



3. Enter exact user name to find or edit users already modified before



4. Assign permissions and ignore default ones.



image
image

√: API tested
√√: API tested + tested in the system
×: not complete

### TODO:

- User without permission to add annotation could still try to add, which may lead to an error. The add button should be disabled.
- Now there is no convenient way to combine per object permissions.

File: debug.log [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[1018/101412.934:ERROR:directory_reader_win.cc(43)] FindFirstFile: ϵͳ�Ҳ���ָ����·���� (0x3)
```
Comment (hunterhector) [inline debug.log:1]: Please remove these

**Comment by hunterhector** [comment]
Closing this temporarily for database cleaning.

**Pull request closed without merging**